### PR TITLE
String to list casting

### PR DESCRIPTION
**Pull request by LindsayWray**
Casts a stringified list to list structure.
Works for all types except structs and maps (yet!)

##### Simple lists:
```
SELECT CAST('[12,13,14]' AS INT[]);
----
[12, 13, 14]  
```

##### And also nested lists:
```
SELECT CAST('[[ [🦆, 🦆, 🦆]], [[duck, db, 🦆], [🦆]], [[🦆, duck, db]]]' AS VARCHAR[][][]);
-----
[[[🦆, 🦆, 🦆]], [[duck, db, 🦆], [🦆]], [[🦆, duck, db]]]
```

#####  Consecutive casting is possible:
```
SELECT CAST(list_value(date)::VARCHAR AS DATE[]) FROM table;
```

#####  And casting within a where clause:
```
SELECT col FROM tbl WHERE LEN(cast(col as int[])) < 4;
```

#####  Redundant whitespace gets filtered out and NULL values are nullified in the list element 
```
SELECT CAST('[[12,13,   14], 	  NULL,   [2,  NULL, 	1]]' AS INT[][]);
```

## The workings:
Each string in the source vector is parsed and split into a list of strings, which results in a new (intermediate) flatvector of strings. Usually this new vector will exceed the source vector’s size due to the splitting (and may be larger than 1024), so new space needs to be allocated. 

Recursively this new vector will be the source vector in its turn. Once enough nesting layers have been stripped, the final casting happens of the varchar list to the list’s target type. This utilises the pre-existing casting functionality. 

For each step, the BindListToListCast is used to determine the next casting function in the casting chain, which is passed using the CastParameters. The (grand)child of the result vector eventually ends the chain by setting the base types, so the resulting vector consists of the target type’s (nested) lists.

### CODE REVIEW

File: src/function/cast/default_casts.cpp [hunk 1]
```diff
@@ -111,7 +111,7 @@ BoundCastInfo DefaultCasts::GetDefaultCastFunction(BindCastInput &input, const L
 	case LogicalTypeId::INTERVAL:
 		return IntervalCastSwitch(input, source, target);
 	case LogicalTypeId::JSON:
-	case LogicalTypeId::VARCHAR:
+	case LogicalTypeId::VARCHAR: // my case
```
Comment (Mytherin) [inline src/function/cast/default_casts.cpp:114]: nit: could you remove the comment?

File: src/function/cast/string_cast.cpp [hunk 2]
```diff
@@ -112,9 +112,100 @@ static BoundCastInfo VectorStringCastNumericSwitch(BindCastInput &input, const L
...
+			idx = sel->get_index(i);
+		}
+		if (!source_mask.RowIsValid(idx)) {
+			continue;
+		}
+		total_list_size += VectorStringifiedListParser::CountParts(source_data[idx]);
+	}
+
+	Vector varchar_vector(LogicalType::VARCHAR, total_list_size); // intermediate cast vector
+	if (total_list_size > count) {
```
Comment (Mytherin) [inline src/function/cast/string_cast.cpp:131]: Could you try adding this piece of code to `Vector::Initialize` and see if that resolves the problem without needing the second resize?

```cpp

	if (capacity > STANDARD_VECTOR_SIZE) {
		validity.Resize(STANDARD_VECTOR_SIZE, capacity);
	}
```

File: test/sql/cast/string_to_struct_cast.test [hunk 3]
```diff
@@ -0,0 +1,463 @@
+# name: test/sql/cast/string_to_struct_cast.test
+# description: cast strings into (nested) list structures
+# group: [cast]
+
+# statement ok
+# PRAGMA enable_verification
```
Comment (Mytherin) [inline test/sql/cast/string_to_struct_cast.test:6]: Could you have a look at fixing this? In general all our tests should run with `enable_verification`

File: src/function/cast/string_cast.cpp [hunk 4]
```diff
@@ -142,6 +233,8 @@ BoundCastInfo DefaultCasts::StringCastSwitch(BindCastInput &input, const Logical
 	case LogicalTypeId::VARCHAR:
 	case LogicalTypeId::JSON:
 		return &DefaultCasts::ReinterpretCast;
+	case LogicalTypeId::LIST: // my case
```
Comment (Mytherin) [inline src/function/cast/string_cast.cpp:236]: nit: could you remove this comment

File: src/function/cast/vector_cast_helpers.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	pos++;
+	while (pos < len && StringUtil::CharacterIsSpace(buf[pos])) {
+		pos++;
+	}
+
+	idx_t start_pos = pos;
+	while (pos < len) {
+		if (buf[pos] == '[') {
+			if (!SkipToClose(++pos, buf, len, ++lvl)) {
```
Comment (Mytherin) [inline src/function/cast/vector_cast_helpers.cpp:84]: What about these two cases:

```sql

["]", "hello", "world"]
[']', "hello", "world"]
```

File: src/function/cast/vector_cast_helpers.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+				return false;
+			}
+		} else if (buf[pos] == '"' || buf[pos] == '\'') {
+			SkipToCloseQuotes(pos, buf, len);
+		} else if (buf[pos] == ',' || buf[pos] == ']') {
+			state.HandleValue(buf, start_pos, pos);
```
Comment (Mytherin) [inline src/function/cast/vector_cast_helpers.cpp:90]: What about the spaces at the end of a value?

File: test/sql/cast/string_to_struct_cast.test [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+query I
+SELECT '[ok,happy]'::MOOD[]
+----
+[ok, happy]
+
+statement error
+SELECT '[nothappy]'::MOOD[]
+
+query I
+SELECT CAST(list_value(date)::VARCHAR AS DATE[]) FROM test_all_types();
```
Comment (Mytherin) [inline test/sql/cast/string_to_struct_cast.test:398]: Maybe you could make this a single list instead using e.g. the `LIST` aggregate?

File: test/sql/cast/string_to_struct_cast.test [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+----
+[5877642-06-25 (BC)]
+[5881580-07-10]
+[NULL]
```
Comment (Mytherin) [inline test/sql/cast/string_to_struct_cast.test:402]: Maybe you can also add some tests for the CSV reader? For example:

```sql
CREATE TABLE int_list(int_list INT[]);

COPY (SELECT [1,2,3]) TO '__TEST_DIR__/int_list.csv';
COPY int_list FROM '__TEST_DIR__/int_list.csv';


```

* Some different types
* Multiple lists
* Nested lists

File: test/sql/cast/string_to_struct_cast.test [hunk 9]
```diff
@@ -0,0 +1,463 @@
+# name: test/sql/cast/string_to_struct_cast.test
```
Comment (Mytherin) [inline test/sql/cast/string_to_struct_cast.test:1]: `string_to_list_cast`

File: test/sql/cast/string_to_struct_cast.test [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+SELECT '[ok,happy]'::MOOD[]
+----
+[ok, happy]
+
+statement error
+SELECT '[nothappy]'::MOOD[]
+
+query I
+SELECT CAST(list_value(date)::VARCHAR AS DATE[]) FROM test_all_types();
+----
```
Comment (Mytherin) [inline test/sql/cast/string_to_struct_cast.test:399]: Empty lists should be castable to any other list (including nested lists):

```sql
SELECT '[]'::INT[];
SELECT '[]'::INT[][];
SELECT '[]'::INT[][][][];

```

This should also work with nested lists:

```sql
SELECT '[[1, 2, 3], [], NULL, [NULL], [4, NULL]]'::INT[][]
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. Some comments below:

**Comment by LindsayWray** [comment]
Thanks @Mytherin ! I think I've now fixed all the things you pointed out!

**Comment by Mytherin** [comment]
Thanks for the fixes! Looks excellent.

**Pull request merged**